packages/datadog-instrumentations/src/apollo.js

```diff
@@ -0,0 +1,100 @@
+const tracingChannel = require('dc-polyfill').tracingChannel
+
+const CHANNELS = {
+  'gateway.request': tracingChannel('apm:apollo:gateway:request'),
+  'gateway.plan': tracingChannel('apm:apollo:gateway:plan'),
+  'gateway.validate': tracingChannel('apm:apollo:gateway:validate'),
+  'gateway.execute': tracingChannel('apm:apollo:gateway:execute'),
+  'gateway.fetch': tracingChannel('apm:apollo:gateway:fetch'),
+  'gateway.postprocessing': tracingChannel('apm:apollo:gateway:postprocessing')
+}
+
+const executorCh = channel('apm:apollo:gateway:request:executor')
+const generalErrorCh = channel('apm:apollo:gateway:general:error')
+
+function wrapExecutor (executor) {
+  return function (...args) {
+    const ctx = { requestContext: args[0], gateway: this }
+    executorCh.publish(ctx)
+    return executor.apply(this, args)
+  }
+}
+
+function wrapApolloGateway (ApolloGateway) {
+  class ApolloGatewayWrapper extends ApolloGateway {
+    constructor (...args) {
+      super(...args)
+      shimmer.wrap(this, 'executor', wrapExecutor)
+    }
+  }
+  return ApolloGatewayWrapper
+}
+
+function wrapRecordExceptions (recordExceptions) {
+  return function wrappedRecordExceptions (...args) {
+    const errors = args[1]
+    // only the last exception in the array of exceptions will be reported on the span,
+    // this is mimicking apollo-gateways internal instrumentation
+    // TODO: should we consider a mechanism to report all exceptions? since this method aggregates all exceptions
+    // where as a span can only have one exception set on it at a time
+    generalErrorCh.publish({ error: errors[errors.length - 1] })
+    return recordExceptions.apply(this, args)
+  }
+}
+
+function wrapStartActiveSpan (startActiveSpan) {
+  return function (...args) {
+    const firstArg = args[0]
+    const cb = args[args.length - 1]
+    if (typeof firstArg !== 'string' || typeof cb !== 'function') return startActiveSpan.apply(this, args)
+
+    const method = CHANNELS[firstArg]
+    let ctx = {}
+    if (firstArg === 'gateway.fetch') {
+      ctx = { attributes: args[1].attributes }
+    }
+
+    switch (firstArg) {
+      case 'gateway.plan' :
+      case 'gateway.validate': {
+        args[args.length - 1] = function (...callbackArgs) {
+          return method.traceSync(cb, ctx, this, ...callbackArgs)
+        }
+        break
+      }
+      case 'gateway.request':
+      case 'gateway.execute':
+      case 'gateway.postprocessing' :
+      case 'gateway.fetch': {
+        args[args.length - 1] = function (...callbackArgs) {
+          return method.tracePromise(cb, ctx, this, ...callbackArgs)
+        }
+        break
+      }
+    }
+    return startActiveSpan.apply(this, args)
+  }
+}
+
+addHook({ name: '@apollo/gateway', file: 'dist/utilities/opentelemetry.js', versions: ['>=2.3.0'] },
+  (obj) => {
+    const newObj = { ...obj }
+    shimmer.wrap(Object.getPrototypeOf(newObj.tracer), 'startActiveSpan', wrapStartActiveSpan)
```

Comment:
Wouldn't this still wrap the same prototype? Why not just patch newObj directly?